xwiki-pro-macros-api/src/main/java/com/xwiki/macros/tab/internal/TabGroupMacro.java

```diff
@@ -0,0 +1,203 @@
+import org.xwiki.rendering.block.GroupBlock;
+import org.xwiki.rendering.block.ListItemBlock;
+import org.xwiki.rendering.block.RawBlock;
+import org.xwiki.rendering.block.match.MacroBlockMatcher;
+import org.xwiki.rendering.macro.MacroContentParser;
+import org.xwiki.rendering.macro.MacroExecutionException;
+import org.xwiki.rendering.macro.descriptor.DefaultContentDescriptor;
+import org.xwiki.rendering.syntax.Syntax;
+import org.xwiki.rendering.transformation.MacroTransformationContext;
+import org.xwiki.skinx.SkinExtension;
+import org.xwiki.text.StringUtils;
+import org.xwiki.xml.XMLUtils;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xwiki.macros.AbstractProMacro;
+import com.xwiki.macros.tab.macro.TabGroupMacroParameters;
+import com.xwiki.macros.tab.macro.TabMacroParameters;
+
+/**
+ * Tab-group macro: the main container macro for tab macro elements.
+ *
+ * @version $Id: $
+ * @since 1.24.0
+ */
+@Component
+@Named("tab-group")
+@Singleton
+public class TabGroupMacro extends AbstractProMacro<TabGroupMacroParameters>
+{
+    private static final String NAME = "Tab group";
+
+    private static final String DESCRIPTION = "The main macro which group tab macro elements.";
+
+    private static final String CONTENT_DESCRIPTION =
+        "It's expected that the content of this macro are only tab macro. "
+            + "Note all other element will be ignored. Only tab macro element will be rendered.";
+
+    private static final String TRUE = "true";
+
+    private static final String FALSE = "false";
+
+    private static final String ID = "id";
+
+    private static final String TAB_MACRO_ID = "tab";
+
+    private static final String TAB_MACRO_PARAM_LABEL = "label";
+
+    private static final String CSS_PX_UNIT = "px;";
+
+    private static final String BLOCK_PARAM_CLASS = "class";
+
+    private static final String BLOCK_PARAM_ROLE = "role";
+
+    private static final Random RANDOM = new Random();
+
+    @Inject
+    private Provider<XWikiContext> xwikiContextProvider;
+
+    @Inject
+    private MacroContentParser contentParser;
+
+    @Inject
+    @Named("ssrx")
+    private SkinExtension ssfx;
+
+    /**
+     * Create and initialize the descriptor of the macro.
+     */
+    public TabGroupMacro()
+    {
+        super(NAME, DESCRIPTION, new DefaultContentDescriptor(CONTENT_DESCRIPTION, true, Block.LIST_BLOCK_TYPE),
+            TabGroupMacroParameters.class);
+    }
+
+    @Override
+    public boolean supportsInlineMode()
+    {
+        return false;
+    }
+
+    private List<Block> renderView(TabGroupMacroParameters parameters, String content,
+        MacroTransformationContext context) throws MacroExecutionException
+    {
+        ssfx.use("css/tabmacro.css");
+        String macroId = parameters.getId();
+
+        if (StringUtils.isEmpty(macroId)) {
+            // FIXME we should have a better way to generate an ID
+            macroId = "tab-group-" + RANDOM.nextInt();
```

Comment:
Does we have a good solution to ensure that we have a unique ID on a page ?

Reply:
Thanks how to ensure that we use the same instance of `IdGenerator` for all macro of the same page, but a new one for each page ?

Reply:
https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-display/xwiki-platform-display-macro/src/main/java/org/xwiki/rendering/internal/macro/display/DisplayMacro.java#L113-L114

This piece of code suggests that the MacroTransformationContext should contain one. Im not sure how it works, but i reckon that the TransformationContext should represent the page that we are in. Otherwise, maybe we could get our hands on the XDOM of the current  page (where the macro is being executed) which should provide an IdGenerator.